components/core/src/clp/ffi/utils.cpp

```diff
@@ -0,0 +1,230 @@
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <utility>
+
+using std::string;
+using std::string_view;
+
+namespace clp::ffi {
+namespace {
+/*
+ * @param byte
+ * @return Whether the input byte is a valid utf8 continuation byte. A valid utf8 continuation byte
+ * should match 0b10xx_xxxx.
+ */
+[[nodiscard]] auto is_valid_utf8_continuation_byte(uint8_t byte) -> bool;
+
+/**
+ * Appends a single-byte utf8 character into the given string, and escapes it if necessary.
+ * @param character Single-byte utf8 character.
+ * @parma escaped_string Input string where the character(s) are appended to.
+ */
+auto escape_and_append_single_byte_utf8_char(uint8_t character, string& escaped_string) -> void;
+
+/**
+ * Validates whether the given code point is a valid UTF8 encoding with the given length.
+ * The valid range is defined as following:
+ * .---------------------------------------------.
+ * | Length | First Code Point | Last Code Point |
+ * |--------|------------------|-----------------|
+ * | 1 Byte | 0x00             | 0x7F            |
+ * | 2 Byte | 0x80             | 0x7FF           |
+ * | 3 Byte | 0x8FF            | 0xFFFF          |
+ * | 4 Byte | 0x10000          | 0x10FFFF        |
+ * |--------|------------------|-----------------|
+ * @param code_point
+ * @param encoding_length
+ * @return Whether the code point is a valid encoding.
+ */
+[[nodiscard]] auto is_valid_code_point(uint32_t code_point, size_t encoding_length) -> bool;
+
+/**
+ * Updates the code point by applying the payload of the given continuation byte.
+ * @param code_point
+ * @param continuation_byte
+ * @return Updated code point.
+ */
+[[nodiscard]] auto update_code_point(uint32_t code_point, uint8_t continuation_byte) -> uint32_t;
+
+auto is_valid_utf8_continuation_byte(uint8_t byte) -> bool {
+    constexpr uint8_t cContinuationByteMask{0xC0};
+    constexpr uint8_t cValidMaskedContinuationByte{0x80};
+    return (byte & cContinuationByteMask) == cValidMaskedContinuationByte;
+}
+
+auto escape_and_append_single_byte_utf8_char(uint8_t character, string& escaped_string) -> void {
+    switch (character) {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        case 0x08:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('b');
+            break;
+        case 0x09:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('t');
+            break;
+        case 0x0A:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('n');
+            break;
+        case 0x0C:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('f');
+            break;
+        case 0x0D:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('r');
+            break;
+        case 0x22:
+            escaped_string.push_back('\\');
+            escaped_string.push_back('\"');
```

Comment:
Why escaped?

Reply:
Hm? You're escaping a double-quote but that escape is not added to the string, right?

```cpp
    if ('\"' == '"') {
        std::cout << "Escape is unnecessary for double-quotes." << std::endl;
    }
```